gamemodes/terrortown/gamemode/server/sv_shop.lua

```diff
@@ -16,138 +16,147 @@ end
 	if shopFallback == SHOP_DISABLED then return end
 
-	-- it's an item if the arg is an id instead of an ent name
-	local id = args[1]
-
-	if not hook.Run("TTTCanOrderEquipment", ply, id) then return end
+	local is_item = items.IsItem(cls)
 
-	local is_item = items.IsItem(id)
+	-- The item/weapon might not even be part of the current shop
+	if not IsPartOfShop(ply, cls) then return end
 
-	-- we use weapons.GetStored to save time on an unnecessary copy, we will not
-	-- be modifying it
-	local equip_table = not is_item and weapons.GetStored(id) or items.GetStored(id)
+	-- we use weapons.GetStored to save time on an unnecessary copy, we will not be modifying it
+	local equip_table = not is_item and weapons.GetStored(cls) or items.GetStored(cls)
 
-	-- some weapons can only be bought once per player per round, this used to be
-	-- defined in a table here, but is now in the SWEP's table
-	if equip_table and equip_table.limited and ply:HasBought(id) then
-		LANG.Msg(ply, "buy_no_stock", nil, MSG_MSTACK_ROLE)
+	if not equip_table then
+		print(ply, "tried to buy equip that doesn't exists", cls)
 
 		return
 	end
 
-	local credits
-
-	if equip_table then
-		-- weapon whitelist check
-		if not table.HasValue(equip_table.CanBuy, subrole) or equip_table.notBuyable then
-			print(ply, "tried to buy equip his subrole is not permitted to buy")
-
-			return
-		end
-
-		-- if we have a pending order because we are in a confined space, don't
-		-- start a new one
-		if HasPendingOrder(ply) then
-			LANG.Msg(ply, "buy_pending", nil, MSG_MSTACK_ROLE)
-
-			return
-		end
+	-- Check for minimum players, global limit, team limit and player limit
+	if not EquipmentIsBuyable(equip_table, ply) then return end
 
-		-- the item is just buyable if there is a special amount of players
-		if not EquipmentIsBuyable(equip_table, ply:GetTeam()) then return end
+	-- still support old items
+	local idOrCls = (is_item and equip_table.oldId or nil) or cls
 
-		credits = equip_table.credits or 1
+	local credits = equip_table.credits or 1
 
-		if credits > ply:GetCredits() then
-			print(ply, "tried to buy item/weapon, but didn't had enough credits")
+	if credits > ply:GetCredits() then
+		print(ply, "tried to buy item/weapon, but didn't have enough credits")
 
-			return
-		end
+		return
+	end
 
-		if GetGlobalInt("ttt2_random_shops") > 0 and RANDOMSHOP[ply] and #RANDOMSHOP[ply] > 0 then
-			local key = false
+	-- keep compatibility with old addons
+	if not hook.Run("TTTCanOrderEquipment", ply, idOrCls, is_item) then return end
 
-			for i = 1, #RANDOMSHOP[ply] do
-				if RANDOMSHOP[ply][i].id ~= id then continue end
+	-- add our own hook with more consistent class parameter and some more information
+	local allow, ignoreCost, message = hook.Run("TTT2CanOrderEquipment", ply, cls, is_item, credits)
 
-				key = true
+	if message then
+		LANG.Msg(ply, message, nil, MSG_MSTACK_ROLE)
+	end
 
-				break
-			end
+	if allow == false then
```

Comment:
this can be simplified into one line (because there isn't any necessary information / no special return like true / false).